libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/EnhancedLocationEngineDecorator.java

```diff
@@ -0,0 +1,185 @@
+
+import timber.log.Timber;
+
+class EnhancedLocationEngineDecorator implements LocationEngine {
+  private static final long UPDATE_INTERVAL_IN_MILLISECONDS = 1000;
+  private static final long FASTEST_UPDATE_INTERVAL_IN_MILLISECONDS = 500;
+  private final LocationEngine locationEngine;
+  private final MapboxNavigator mapboxNavigator;
+  private final LocationEngineCallback<LocationEngineResult> callback = new CurrentLocationEngineCallback(this);
+  private ScheduledFuture future;
+  private final ScheduledExecutorService executorService;
+  private Location rawLocation;
+  private Location enhancedLocation;
+
+  EnhancedLocationEngineDecorator(@NonNull LocationEngine locationEngine,
+                                  @NonNull MapboxNavigator mapboxNavigator,
+                                  @NonNull ScheduledExecutorService executorService) {
+    this.locationEngine = locationEngine;
+    this.mapboxNavigator = mapboxNavigator;
+    this.executorService = executorService;
+    onFreeDrive();
+  }
+
+  @Override
+  public void getLastLocation(@NonNull LocationEngineCallback<LocationEngineResult> callback) throws SecurityException {
+    locationEngine.getLastLocation(callback);
+  }
+
+  @Override
+  public void requestLocationUpdates(@NonNull LocationEngineRequest request,
+                                     @NonNull final LocationEngineCallback<LocationEngineResult> callback,
+                                     @Nullable Looper looper) throws SecurityException {
+    locationEngine.requestLocationUpdates(request, callback, null);
+  }
+
+  @Override
+  public void requestLocationUpdates(@NonNull LocationEngineRequest request,
+                                     PendingIntent pendingIntent) throws SecurityException {
+    locationEngine.requestLocationUpdates(request, pendingIntent);
+  }
+
+  @Override
+  public void removeLocationUpdates(@NonNull LocationEngineCallback<LocationEngineResult> callback) {
+    locationEngine.removeLocationUpdates(callback);
+  }
+
+  @Override
+  public void removeLocationUpdates(PendingIntent pendingIntent) {
+    locationEngine.removeLocationUpdates(pendingIntent);
+  }
+
+  void onActiveGuidance() {
+    removeLocationUpdates(callback);
+    future.cancel(false);
+    future = null;
+  }
+
+  void onFreeDrive() {
+    requestLocationUpdates(obtainLocationEngineRequest(), callback, null);
+    final Handler handler = new Handler(Looper.getMainLooper());
+    future = executorService.scheduleAtFixedRate(new Runnable() {
+      @Override
+      public void run() {
+        if (rawLocation != null) {
+          enhancedLocation = getLocation(new Date(), 0, rawLocation);
+          handler.post(new Runnable() {
+            @Override
+            public void run() {
+              callback.onSuccess(LocationEngineResult.create(enhancedLocation));
+            }
+          });
+        }
+      }
+    }, 1500, 1000, TimeUnit.MILLISECONDS);
+  }
+
+  void onNavigationKilled() {
```

Comment:
Should we manually shutdown the executor here?

Reply:
Yeah, that's a great idea, after cancelling the future task though.

Fixing.